### ISSUE TITLE
Epdfinfo quits when using poppler 0.85

### ISSUE THREAD

**Issue by dalanicolai** [politza/pdf-tools#548]
Today I installed the new poppler, that comes with libpoppler.so.95. However after rebuilding the epdfinfo server, the server immediately quits when opening a book and prints "Error while processing tq callback: The epdfinfo server quit unexpectedly" in the message buffer.

When building again with libpoppler.so.84, epdfinfo succesfully opens books again.

UNRELATED COMMENT
I am almost sure this is not the correct place for this comment, so sorry for that, but poppler supports arrow annotations now (e.g. Okular supports line annotations with arrowheads now). It would be amazing if these could be implemented in pdf-tools, even more if there could be a persistent annotation mode, so that drawing multiple arrows in a row would be possible (without need for reselecting arrow annotation "mode"). Unfortunately I do not know enough C yet to implement it myself. So for now I will just use Okular if when required :)

**Comment by wakatara** [comment]
I have the same problem as @dalani but with the poppler 0.86.1 that homebrew just updated on my OSX 10.15.3. 

PDFTools loaded the text of the pdf rather than rendered, then tried to recompile the pdftools (successfully) but the epdfinfo server keeps quitting makinf the tool not work (sadly, because I am in the middle of a textbook.)... =p 

Any ideas on how to resolve this other than droppig down to a previous version of poppler (assuming that is the culprit.).

(this is on emacs 26.3 in case that is key).

**Comment by wakatara** [comment]
(please note I am using pdf-tools with the excellent emacs interleave-mode to get side-by-side notes and pdf in emacs... =] ).

**Comment by politza** [comment]
It's lilely a problem with your system configuration.  Try emacs.stackexchange.com .

**Comment by wakatara** [comment]
Sorry, I should have noted that I was using pdf-tools fine with emacs interleave mode fine before homebrew upgraded poppler and another pdf-tools depdency. Will do some googling, but my bet would be it is something about poppler and pdf-tools' epdfinfo server not playing nicely. Will ask on SO and report back if anything comes of it here. 

Thanks for working on such an really useful tool. It makes my scientific and te4xtbook reading (with interleave mode. =] ) much much more productive.

**Comment by smallzhan** [comment]
same issue with poppler 0.86.1 here, I've downgrade poppler to 0.85.0, it works nicely.

**Comment by politza** [comment]
Try 

echo quit | path/to/epdfinfo

in a shell and post the output of that command.

**Comment by wakatara** [comment]
Hey @politza 

I get a clean OK and . from the command you suggested.

```shell
~/.e/e/pdf-tools-20200214.536 echo quit | /Users/daryl/.emacs.d/elpa/pdf-tools-20200214.536/epdfinfo
OK
.
```

Though the issue is the server crashes when I try to move through a pdf now that poppler 0.86.1 is installed not that is has trouble spinning up/starting. I actually uninstalled the entire pdf-tools package, removed all homebrew packages affecting it, updated homebrew and then installed and recompiled the pdf-tools package  but still seeing the same issue. 

Both with interleave-mode and normal pdf viewing (loading a pdf file) mode, triyng to hit 'n' or anything to move onto further pages results in the epdfinfo server crashing and the page being the only one being able to be seen. :-/  

(I'm going to go look into downgrading poppler down to 0.85 as suggested above to see if that solves the issue.). 

Please let me know how I may help diagnose the issue better for you to help solve the bug (assuming it's not an OSX quirk.). Happy to help out. 

thanks!

**Comment by wakatara** [comment]
Hey @politza (and @smallzhan )

(also, posting instructions for people who may be having same problem on OSX to deprecate to 0.85 poppler to temporarily correct till this or poppler are fixed.).

On OSX, I used `brew uninstall poppler` (which was at 0.86.1) and then used 

```
brew install https://raw.githubusercontent.com/Homebrew/homebrew-core/ae0b8ada9ef4e849087cbd3bbf74e3d29b5241ab/Formula/poppler.rb
```
to install the 0.85 version of poppler.
To keep it fixed on 0.85 till there is a correction, I then used `brew pin poppler` which will prevent updates to it. 

When I launched emacs again, pdf-tools automatically told me it needed a recomplile and went and did that happily (this is a freaking nice feature by the way!).

Both interleave mode and normal pdf loading via the file dialogue now working. 

Def seems to be either an issue with the 0.86.1 poppler library on OSX or the interaction of pdf-tools with it on OSX. 

ciao! 
(and thanks to @smallzhan for mentioning that downgrading worked.).

**Comment by kleinschmidt** [comment]
I noticed what I thought was the same problem after updating to poppler 0.86.1 on archlinux, but it turns out I just needed to do a completely clean build of pdf-tools by removing the package and adding it again, at which point it linked against the correct libpoppler.

**Comment by wakatara** [comment]
@kleinschmidt 

Thanks for the suggestion! I just tried uninstalling pdf-tools (and packages that depended on it, cleaning out all built files.). 

I then unpinned poppler, reinstalled 0.86.1 and had it rebuild pdf-tools from scratch. Sadly, exactly the same problem with epdfinfo crashing so I believe it is probably similarly related but does not solve the epdfinfo crashing problem on OSX (sad 🐼  ).

Besides doing a clean rebuild was there anything else you did that might help me diagnose why pdf-tools, osx, and poppler appear to not be playing well? 

I've downgraded to poppler 0.85 and everything is working fine again (as per @smallzhan 's pointer above.).

I might try and see whether I have the same problem on my POP OS system later though I believe this appears to be something unique to the interaction between osx, poppler 0.86.1, and pdf-tools sadly. (to be know what changed between 0.85 and 0.86.1 in poppler?).

**Comment by wakatara** [comment]
Interestingly, going over and looking at the poppler gitlab there is an isue filed 4 hours ago on [osx poppler segfault issue](https://gitlab.freedesktop.org/poppler/poppler/issues/893) that looks very relevant. Will keep an eye on it though imagine it's exactly what's causing the issue here.

**Comment by cfraizer** [comment]
I have the same problem with:
emacs 26.3
epdfinfo is using:
	/usr/local/opt/poppler/lib/libpoppler-glib.8.dylib (compatibility version 8.0.0, current version 8.15.0)
	/usr/local/opt/poppler/lib/libpoppler.97.dylib (compatibility version 97.0.0, current version 97.0.0)

The "quit" command works just fine.

After trying the commands in order, I found that `(pdf-info-pagelinks 10 "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")`
caused this backtrace:

```
Debugger entered--Lisp error: (error "The epdfinfo server quit unexpectedly.")
  signal(error ("The epdfinfo server quit unexpectedly."))
  error("The epdfinfo server quit unexpectedly.")
  (progn (error "The epdfinfo server quit unexpectedly."))
  (if (and (not done) (not (eq (process-status (pdf-info-process)) (quote run))) (not (eq cmd (quote quit)))) (progn (error "The epdfinfo server quit unexpectedly.")))
  (if pdf-info-asynchronous nil (while (and (not done) (eq (process-status (pdf-info-process)) (quote run))) (accept-process-output (pdf-info-process) 0.01)) (if (and (not done) (not (eq (process-status (pdf-info-process)) (quote run))) (not (eq cmd (quote quit)))) (progn (error "The epdfinfo server quit unexpectedly."))) (cond ((null status) response) ((eq status (quote error)) (error "epdfinfo: %s" response)) ((eq status (quote interrupted)) (error "epdfinfo: Command was interrupted")) (t (error "internal error: invalid response status"))))
  (let* ((query (concat (mapconcat (quote pdf-info-query--escape) (cons cmd args) ":") "\n")) (callback (function (lambda (closure response) (let* ((result ...) (status ...)) (progn (pdf-info-query--log response) (let* ... ...)))))) response status done (closure (or pdf-info-asynchronous (function (lambda (s r) (setq status s response r done t)))))) (pdf-info-query--log query t) (tq-enqueue pdf-info--queue query "^\\.\n" closure callback) (if pdf-info-asynchronous nil (while (and (not done) (eq (process-status (pdf-info-process)) (quote run))) (accept-process-output (pdf-info-process) 0.01)) (if (and (not done) (not (eq (process-status (pdf-info-process)) (quote run))) (not (eq cmd (quote quit)))) (progn (error "The epdfinfo server quit unexpectedly."))) (cond ((null status) response) ((eq status (quote error)) (error "epdfinfo: %s" response)) ((eq status (quote interrupted)) (error "epdfinfo: Command was interrupted")) (t (error "internal error: invalid response status")))))
  pdf-info-query(pagelinks "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf" 10)
  pdf-info-pagelinks(10 "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
  eval((pdf-info-pagelinks 10 "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf") nil)
  #f(compiled-function (exp &optional insert-value no-truncate char-print-limit) "Evaluate EXP and print value in the echo area.\nWhen called interactively, read an Emacs Lisp expression and\nevaluate it.  Value is also consed on to front of the variable\n`values'.  Optional argument INSERT-VALUE non-nil (interactively,\nwith a non `-' prefix argument) means insert the result into the\ncurrent buffer instead of printing it in the echo area.\n\nNormally, this function truncates long output according to the\nvalue of the variables `eval-expression-print-length' and\n`eval-expression-print-level'.  When NO-TRUNCATE is\nnon-nil (interactively, with a prefix argument of zero), however,\nthere is no such truncation.\n\nIf the resulting value is an integer, and CHAR-PRINT-LIMIT is\nnon-nil (interactively, unless given a non-zero prefix argument)\nit will be printed in several additional formats (octal,\nhexadecimal, and character).  The character format is only used\nif the value is below CHAR-PRINT-LIMIT (interactively, if the\nprefix argument is -1 or the value doesn't exceed\n`eval-expression-print-maximum-character').\n\nRuns the hook `eval-expression-minibuffer-setup-hook' on entering the\nminibuffer.\n\nIf `eval-expression-debug-on-error' is non-nil, which is the default,\nthis command arranges for all errors to enter the debugger." (interactive #f(compiled-function () #<bytecode 0x5dab87a5>)) #<bytecode 0x400c0aaf>)((pdf-info-pagelinks 10 "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf") nil nil 127)
  apply(#f(compiled-function (exp &optional insert-value no-truncate char-print-limit) "Evaluate EXP and print value in the echo area.\nWhen called interactively, read an Emacs Lisp expression and\nevaluate it.  Value is also consed on to front of the variable\n`values'.  Optional argument INSERT-VALUE non-nil (interactively,\nwith a non `-' prefix argument) means insert the result into the\ncurrent buffer instead of printing it in the echo area.\n\nNormally, this function truncates long output according to the\nvalue of the variables `eval-expression-print-length' and\n`eval-expression-print-level'.  When NO-TRUNCATE is\nnon-nil (interactively, with a prefix argument of zero), however,\nthere is no such truncation.\n\nIf the resulting value is an integer, and CHAR-PRINT-LIMIT is\nnon-nil (interactively, unless given a non-zero prefix argument)\nit will be printed in several additional formats (octal,\nhexadecimal, and character).  The character format is only used\nif the value is below CHAR-PRINT-LIMIT (interactively, if the\nprefix argument is -1 or the value doesn't exceed\n`eval-expression-print-maximum-character').\n\nRuns the hook `eval-expression-minibuffer-setup-hook' on entering the\nminibuffer.\n\nIf `eval-expression-debug-on-error' is non-nil, which is the default,\nthis command arranges for all errors to enter the debugger." (interactive #f(compiled-function () #<bytecode 0x42617179>)) #<bytecode 0x400c0aaf>) ((pdf-info-pagelinks 10 "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf") nil nil 127))
  eval-expression((pdf-info-pagelinks 10 "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf") nil nil 127)
  funcall-interactively(eval-expression (pdf-info-pagelinks 10 "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf") nil nil 127)
  #<subr call-interactively>(eval-expression nil nil)
  apply(#<subr call-interactively> eval-expression (nil nil))
  call-interactively@ido-cr+-record-current-command(#<subr call-interactively> eval-expression nil nil)
  apply(call-interactively@ido-cr+-record-current-command #<subr call-interactively> (eval-expression nil nil))
  call-interactively(eval-expression nil nil)
  command-execute(eval-expression)
```

The only text in the `␣*epdfinfo*` buffer is: `Process epdfinfo segmentation fault: 11`.

Commands that work:
```
(pdf-info-boundingbox 10 "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
(pdf-info-charlayout 10 nil "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
(pdf-info-encrypted-p "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
(pdf-info-features)
    → (case-sensitive-search writable-annotations markup-annotations)
(pdf-info-getannots (cons 1 200) "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
(pdf-info-getannot 'annot-2-3 "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
(pdf-info-getattachments nil "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
(pdf-info-getoptions "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
    → (:render/background #000000 :render/foreground #000000 :render/printed nil :render/usecolors nil)
(pdf-info-getselection 10 (pdf-info-boundingbox 10 "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf") 'line "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
(pdf-info-gettext 10 (pdf-info-boundingbox 10 "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf") 'line "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
(pdf-info-interrupt)
(pdf-info-kill)
(pdf-info-metadata "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
(pdf-info-search-string "interpretation" (cons 1 100) "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
(pdf-info-search-regexp "[ij]nterpretation" (cons 1 100) t "/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf")
```

**Comment by cfraizer** [comment]
I tried something similar with another PDF.

`epdfinfo` crashes whenever `pdf-info-pagelinks` is invoked.

Also, from the command line, I invoked `epdfinfo` and fed it commands like
```
search-string:/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf:1:200:interpretation:0
OK
177:interpretation:and the interpretation of this as a tree (as in Figure 2.6).:0.275964 0.706279 0.423842 0.728549
.
pagelinks:/Users/colinf/Dropbox/AAA_Books/Computer Science/S/Structure and Interpretation of Computer Programs-2nd ed.pdf:10
Segmentation fault: 11
```

**Comment by cfraizer** [comment]
This [patch](https://gitlab.freedesktop.org/poppler/poppler/issues/893#note_436769) fixed the issues for me.

**Comment by CsBigDataHub** [comment]
@cfraizer ,

Thanks for the update . Do you know when the next Homebrew release will go out.

**Comment by wakatara** [comment]
@cfraizer and @CsBigDataHub 
Homebrew has already been updated. I just upgraded poppler, kicked into emacs, the pdf-tools recompiled automagically, and am happily using pdf-tools again with the 0.86.1_1 version in interleave mode. 

Looks all good. I think this one is good to close.

**Comment by CsBigDataHub** [comment]
@wakatara ,
Thanks, Yes its seems to be working now. Upgraded poppler through Homebrew and emacs rebuild epdfinfo.

**Comment by dalanicolai** [comment]
Hi! This problem still occurs on linux. However I would say it became a critical problem now since Fedora as well as Ubuntu upgraded to new poppler versions, so that on both my Fedora and Ubuntu installations this problem started occuring (until now I avoided it by using the default "older" Fedora/Ubuntu poppler versions). When I open a pdf file it first shows correctly but also the server quits unexpectedly. I don't know exactly which version of libpoppler is used to build the epdfinfo server but I find that the following files/packages on my systems:

Fedora:
poppler 0.84
libpoppler.so.90 & 94 & 95 (in /usr/share/lib64)

Ubuntu:
poppler 0.86
libpoppler.so.90 & 97

On both systems I removed the pdf-tools directory so that pdf-tools was reinstalled from melpa and also the epdfinfo server was rebuild. Additionally I tried to downgrade the poppler installations but there were too much dependencies that required the new poppler versions (e.g. latex-xetex). Subsequently I tried using a manually build the server with poppler 0.88 (and libpoppler.so.99), where epdfinfo complained that it could not find poppler-private headers, but I was able to solve it by adding the right path to the headers to the poppler source directory (I can not easily find how I should install these headers to some "system directories" via `make install`). Anyway, would be great if you could advise me about a solution. But anyway I guess this is a critical bug now. Of course I am happy to provide more informatio with your guidance. Here follows a backtrace:
```
Debugger entered--Lisp error: (error "The epdfinfo server quit unexpectedly.")
  signal(error ("The epdfinfo server quit unexpectedly."))
  error("The epdfinfo server quit unexpectedly.")
  pdf-info-query(pagelinks "/mnt/4EEDC07F44412A81/git/Masterthesis/main.pdf" 3)
  pdf-info-pagelinks(3)
  pdf-cache-pagelinks(3)
  pdf-links-hotspots-function(3 (1584 . 2241))
  #f(compiled-function (fn) #<bytecode 0x3732cd1>)(pdf-links-hotspots-function)
  mapcar(#f(compiled-function (fn) #<bytecode 0x3732cd1>) (pdf-annot-hotspot-function pdf-links-hotspots-function pdf-view-text-regions-hotspots-function))
  pdf-view-apply-hotspot-functions(nil 3 (1584 . 2241))
  pdf-view-create-page(3)
  #f(compiled-function (status data) #<bytecode 0x3636d29>)(nil "\211PNG\... some huge byte string here ...\202")
  #f(compiled-function (closure response) #<bytecode 0x3636d35>)(#f(compiled-function (status data) #<bytecode 0x3636d29>) "OK\n/tmp/epdfimwzm4j\n.\n")
  ad-Advice-tq-process-buffer(#f(compiled-function (tq) "Check TQ's buffer for the regexp at the head of the queue." #<bytecode 0x41b5189>) (((nil "^\\.\n" #f(compiled-function (s r) #<bytecode 0x3732d8d>) . #f(compiled-function (closure response) #<bytecode 0x3732d6d>))) #<process epdfinfo> . #<killed buffer>))
  apply(ad-Advice-tq-process-buffer #f(compiled-function (tq) "Check TQ's buffer for the regexp at the head of the queue." #<bytecode 0x41b5189>) (((nil "^\\.\n" #f(compiled-function (s r) #<bytecode 0x3732d8d>) . #f(compiled-function (closure response) #<bytecode 0x3732d6d>))) #<process epdfinfo> . #<killed buffer>))
  tq-process-buffer((((nil "^\\.\n" #f(compiled-function (s r) #<bytecode 0x3732d8d>) . #f(compiled-function (closure response) #<bytecode 0x3732d6d>))) #<process epdfinfo> . #<killed buffer>))
  tq-filter((((nil "^\\.\n" #f(compiled-function (s r) #<bytecode 0x3732d8d>) . #f(compiled-function (closure response) #<bytecode 0x3732d6d>))) #<process epdfinfo> . #<killed buffer>) "OK\n/tmp/epdfimwzm4j\n.\n")
  #f(compiled-function (proc string) #<bytecode 0x148b9e9>)(#<process epdfinfo> "OK\n/tmp/epdfimwzm4j\n.\n")
```

**Comment by politza** [comment]
Can you provide a PDF file where this crash occurs ?  I can not reproduce in using poppler 0.88.0-1 on Archlinux.

**Comment by dalanicolai** [comment]
Yes sure...
[testfile.pdf](https://github.com/politza/pdf-tools/files/4637065/testfile.pdf)

Indeed I noticed now many files are working opening correctly but not my files created from latex (using pdflatex)

**Comment by dalanicolai** [comment]
Do you know how I can tell the ./configure where to find the poppler-private headers (when buidling poppler manually). I somehow managed to do it before when building poppler 0.88. I created a note that said I had to add the right folder to CPPFLAGS in my /usr/share/config.site file. However now I try to build an older version (poppler 0.73) and this trick does not seem to work (maybe I wrote down a wrong note, although I'm almost sure this was the solution). Anyway... if I know how to include the path to the private header then I can test again with poppler 0.73.

**Comment by dalanicolai** [comment]
I found now how to add the poppler-private headers via the -DENABLE_UNSTABLE_API_ABI_HEADERS=on flag and I was able to build poppler 0.73 and rebuild the epdfinfo server with it. The files open correctly now.

**Comment by dalanicolai** [comment]
Okay so finally, I tried again to build the server with poppler version 0.88 and indeed the files open correctly. So I think in my previous comment, although I managed to build poppler 0.88 with the private headers, I think I only configured the epdfinfo server but actually forgot to run the `./autobuild -i ~/.emacs.d/elpa/26.3/dev
elop/pdf-tools-20200512.1524/` command (where I skip the installation of system packages using Ctrl-C). So in summary the epdfinfo server seems to work fine with poppler 0.88, but not with the current default system packages in Fedora and Ubuntu. Anyway I guess the problem will solve itself when Fedora and Ubuntnu start to ship the newer poppler versions. Anyway, thanks for your reply and maybe this issue can be closed now. Or maybe leave it open until Fedora and Ubuntu upgrade their packages system packages.
